File: AED_II/AED_TP/TP3/Questao5/Jogador.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Objects;

class Celula{
    public Jogador jogador;
    public Celula prox;
    public Celula(){
        this(null);
    }

    public Celula (Jogador jogador){
        this.jogador = jogador;
        this.prox = null;
    }
    
}

class Lista {
    private Celula primeiro,ultimo;
    public Lista() {
        primeiro = new Celula();
        ultimo = primeiro;
    }

    public void inserirFim(Jogador jogador) throws Exception {
        ultimo.prox = new Celula(jogador);
        ultimo = ultimo.prox;
    }

    public void inserirInicio(Jogador jogador) throws Exception {
        Celula tmp = new Celula(jogador);
        tmp.prox = primeiro.prox;
        primeiro.prox = tmp;
        if(primeiro == ultimo){
            ultimo = tmp;
        }
        tmp = null;

    }

    public void inserir(Jogador jogador, int pos) throws Exception {
        int tamanho = tamanho();
        if (pos < 0 || pos > tamanho){ 
            throw new Exception();
        } else if (pos == 0){ 
            inserirInicio(jogador);
        } else if (pos == tamanho){ 
            inserirFim(jogador);
        } else {
        Celula i = primeiro;
        for(int j = 0; j < pos; j++, i = i.prox); 
        Celula tmp = new Celula(jogador);
        tmp.prox = i.prox;
        i.prox = tmp;
        tmp = i = null;
        }
    }

    public Jogador removerInicio() throws Exception {
        if(primeiro == ultimo)
            throw new Exception();
        
        Celula tmp = primeiro;
        primeiro = primeiro.prox;
        Jogador jogador = primeiro.jogador;
        tmp.prox = null;
        tmp = null;
        return jogador;
    }

    public Jogador remover(int pos) throws Exception {
        int tamanho = tamanho();
        Jogador jogador;
        if (primeiro == ultimo || pos < 0 || pos >= tamanho){ throw new Exception();
        } else if (pos == 0) { jogador = removerInicio();
        } else if (pos == tamanho - 1){ jogador = removerFim();
        } else {
        Celula i = primeiro;
        for(int j = 0; j < pos; j++, i = i.prox);
        Celula tmp = i.prox;
        jogador = tmp.jogador; i.prox = tmp.prox;
        tmp.prox = null; i = tmp = null;
        }
        return jogador;
    }

    public Jogador removerFim() throws Exception {
    if (primeiro == ultimo)
        throw new Exception();
    Celula i;
    for(i = primeiro; i.prox != ultimo; i = i.prox);
    Jogador jogador = ultimo.jogador;
    ultimo = i;
    i = ultimo.prox = null;
    return jogador;
    }

    public int tamanho() {
      int tamanho = 0; 
      for(Celula i = primeiro; i != ultimo; i = i.prox, tamanho++);
      return tamanho;
   }

    public void imprimir() {
    int posicao = 0;
        for (Celula i = primeiro.prox; i != null; i = i.prox) {
        if (i.jogador != null) {
            i.jogador.imprimirJogador(posicao);
        }
        posicao++;
    }

    }

    
}

public class Jogador {
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;

    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        this.id = id;
        this.nome = nome;
        this.altura = altura;
        this.peso = peso;
        this.universidade = universidade;
        this.anoNascimento = anoNascimento;
        this.cidadeNascimento = cidadeNascimento;
        this.estadoNascimento = estadoNascimento;
    }

    public Jogador(){

    }

    public void setId(int id){
        this.id = id;
    }

    public int getId(){
        return id;
    }

    public void setNome(String nome){
        this.nome = nome;
    }

    public String getNome(){
        return nome;
    }

    public void setAltura(int altura){
        this.altura = altura;
    }

    public int getAltura(){
        return altura;
    }

    public void setPeso(int peso){
        this.peso = peso;
    }

    public int getPeso(){
        return peso;
    }

    public void setUniversidade(String universidade){
        this.universidade = universidade;
    }

    public String getUniversidade(){
        return universidade;
    }

    public void setAnoNascimento(int anoNascimento){
        this.anoNascimento = anoNascimento;
    }

    public int getAnoNascimento(){
        return anoNascimento;
    }

    public void setCidadeNascimento(String cidadeNascimento){
        this.cidadeNascimento = cidadeNascimento;
    }

    public String getCidadeNascimento(){
        return cidadeNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento){
        this.estadoNascimento = estadoNascimento;
    }

    public String getEstadoNascimento(){
        return estadoNascimento;
    }

    public Jogador clone(){
        Jogador novoJogador = new Jogador(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento);
        return novoJogador;
    }

    public void imprimirJogador(int pos){
        System.out.println("[" + pos + "]" + " ## " + this.nome + " ## " + this.altura + " ## " + this.peso + " ## " + this.anoNascimento + " ## " + this.universidade + " ## " + this.cidadeNascimento + " ## " + this.estadoNascimento + " ##");
    }
    

    public void ler(BufferedReader br, String linha) throws IOException {
        if (linha != null && !linha.equals("FIM")) {
            String[] valores = linha.split(",", -1);

            // Verifica se existem valores suficientes no array antes de acessá-los
            if (valores.length >= 8) {
                // Certifique-se de tratar exceções ao fazer parsing dos valores
                try {
                    this.id = Integer.parseInt(valores[0]);
                    this.nome = (valores[1].isEmpty()) ? "nao informado" : valores[1];
                    this.altura = Integer.parseInt(valores[2]);
                    this.peso = Integer.parseInt(valores[3]);
                    this.universidade = (valores[4].isEmpty()) ? "nao informado" : valores[4];
                    this.anoNascimento = Integer.parseInt(valores[5]);
                    this.cidadeNascimento = (valores[6].isEmpty()) ? "nao informado" : valores[6];
                    this.estadoNascimento = (valores[7].isEmpty()) ? "nao informado" : valores[7];
                } catch (NumberFormatException e) {
                    
                }
            }
        }
    }

    private static void swap(int index1, int index2, Jogador[] jogadores) {
            Jogador temp = jogadores[index1];
            jogadores[index1] = jogadores[index2];
            jogadores[index2] = temp;
    }
    // ...
    
    public static void main(String[] args) {
        String nomeArquivo = "/tmp/players.csv"; 
        HashMap<Integer, Jogador> jogadoresPorId = new HashMap<>();
        Lista listaJogadores = new Lista();
        int[] ID = new int[3923];
        int a = 0;
        Jogador[] jogadores = new Jogador[3923];

        try (BufferedReader br = new BufferedReader(new FileReader(nomeArquivo))) {
            br.readLine(); // Ignora a primeira linha (cabeçalho)

            // Armazena todos os jogadores no HashMap
            String linha;
            while ((linha = br.readLine()) != null) {
                Jogador jogador = new Jogador();
                jogador.ler(br, linha);
                jogadoresPorId.put(jogador.getId(), jogador);
            }

            BufferedReader entradaPadrao = new BufferedReader(new InputStreamReader(System.in));
            
        // Ler IDs da entrada padrão
        while (true) {
            try {
                String entrada = entradaPadrao.readLine();

                // Using Objects.equals() to prevent NullPointerException
                if ("FIM".equals(entrada)) {
                    break;
                }

                int idProcurado = Integer.parseInt(entrada);
                ID[a] = idProcurado;
                a++;

            } catch (NumberFormatException e) {
                // Handle exception as per your requirement
            }
        }

        for (int j = 0; j < a; j++) {
            try{
                listaJogadores.inserirFim(jogadoresPorId.get(ID[j]));
            } catch (Exception e) {

            }
        }
       
        String entrada = entradaPadrao.readLine();
        int numRegistros = 0;
        if (entrada != null && !entrada.isEmpty()) {
            try {
                numRegistros = Integer.parseInt(entrada);
                // Processamento adicional
            } catch (NumberFormatException e) {
            }
        }
        for (int i = 0; i < numRegistros; i++){
                String entrada3 = entradaPadrao.readLine();
                if (entrada3 != null && !entrada3.isEmpty()){
                    if(entrada3.charAt(0) == 'I'){
                        if(entrada3.charAt(1) == 'I'){
                            String numeros = entrada3.replaceAll("[^\\d]", ""); // Remove tudo que não é dígito
                            int id = Integer.parseInt(numeros);
                            try{
                            listaJogadores.inserirInicio(jogadoresPorId.get(id));
                            }catch (Exception e) {}
                        }else if(entrada3.charAt(1) == 'F'){
                            String numeros = entrada3.replaceAll("[^\\d]", ""); // Remove tudo que não é dígito
                            int id = Integer.parseInt(numeros);
                            try{
                            listaJogadores.inserirFim(jogadoresPorId.get(id));
                            }catch (Exception e) {}
                        }else if(entrada3.charAt(1) == '*'){
                            String[] partes = entrada3.split("\\s+"); // Divide a string com base em um ou mais espaços
                            int posicao = Integer.parseInt(partes[1]);
                            int id = Integer.parseInt(partes[2]);
                            try{
                            listaJogadores.inserir(jogadoresPorId.get(id), posicao);
                            }catch (Exception e) {}
                        }
                    }else if(entrada3.charAt(0) == 'R'){
                        if(entrada3.charAt(1) == 'I'){
                            try{
                            Jogador removido = listaJogadores.removerInicio();
                            System.out.println("(R) " + removido.nome);
                            }catch (Exception e) {}
                        }else if(entrada3.charAt(1) == 'F'){
                            try{
                            Jogador removido = listaJogadores.removerFim();
                            System.out.println("(R) " + removido.nome);
                            }catch (Exception e) {}
                        }else if(entrada3.charAt(1) == '*'){
                            String numeros = entrada3.replaceAll("[^\\d]", ""); // Remove tudo que não é dígito
                            int posicao = Integer.parseInt(numeros);
                            try{
                            Jogador removido = listaJogadores.remover(posicao);
                            System.out.println("(R) " + removido.nome);
                            }catch (Exception e) {}
                        }
                    }
                }
        }
       
        
        
        listaJogadores.imprimir();


        } catch (IOException e) {
            
        }
    }
}

File: 2_Periodo/AED_II/AED_TP/TP3/Questao7/jogador.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef struct Jogador {
    int id;
    char nome[250];
    int altura;
    int peso;
    char universidade[250];
    int anoNascimento;
    char cidadeNascimento[250];
    char estadoNascimento[250];
} Jogador;

typedef struct Celula {
    Jogador jogador;
    struct Celula *prox;
} Celula;

typedef struct Fila {
    Celula *primeiro;
    Celula *ultimo;
    int tamanho;
} Fila;

void start(Fila *fila);
void inserir(Jogador jogador, Fila *fila);
Jogador remover(Fila *fila);  
void imprimir(Jogador *jogador);
int lerJogador(FILE* arquivo, Jogador* jogador);

void start(Fila *fila) {
    fila->primeiro = fila->ultimo = NULL;
    fila->tamanho = 0;
}

void inserir(Jogador jogador, Fila *fila) {
    if (fila->tamanho == 5) {
        remover(fila); // Remove o elemento mais antigo se a fila estiver cheia
    }

    Celula *temp = (Celula*) malloc(sizeof(Celula));
    temp->jogador = jogador;
    temp->prox = NULL;

    if (fila->ultimo != NULL) {
        fila->ultimo->prox = temp;
    } else {
        fila->primeiro = temp;
    }

    fila->ultimo = temp;
    fila->tamanho++;
}

Jogador remover(Fila *fila) {
    if (fila->primeiro == NULL) {
        exit(1); // Fila vazia
    }

    Celula *temp = fila->primeiro;
    Jogador removido = temp->jogador;

    fila->primeiro = fila->primeiro->prox;
    if (fila->primeiro == NULL) {
        fila->ultimo = NULL;
    }

    free(temp);
    fila->tamanho--;
    return removido;
}

void imprimir(Jogador *jogador) {
    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##]\n", jogador->id, jogador->nome, jogador->altura, jogador->peso, jogador->anoNascimento, jogador->universidade, jogador->cidadeNascimento, jogador->estadoNascimento);
}

void verificaVazio(char* campo) {
    if (strcmp(campo, " ") == 0 || strcmp(campo, "") == 0) {
        strcpy(campo, "nao informado");
    }
}

char *scan(char **pp, char c)
{
    char *s = *pp, *p;
    p = strchr(*pp, c);
    if (p) *p++ = '\0';
    *pp = p;
    return s;
}

int lerJogador(FILE* arquivo, Jogador* jogador) {
    char linha[1024];
    char *v[8],*pointLin;
    int i;
    if (fgets(linha, sizeof(linha), arquivo) == NULL) {
        return EOF; // Fim do arquivo
    }

    // Dividir a linha em campos
    pointLin = linha;
    for (i = 0; i < 8; i++) v[i] = scan(&pointLin, ',');
    
    jogador->id = atoi(v[0]);
    strcpy(jogador->nome, v[1]);
    jogador->altura = atoi(v[2]);
    jogador->peso = atoi(v[3]);
    strcpy(jogador->universidade, v[4]);
    jogador->anoNascimento = atoi(v[5]);
    strcpy(jogador->cidadeNascimento, v[6]);
    strcpy(jogador->estadoNascimento, v[7]);
    jogador->estadoNascimento[strlen(jogador->estadoNascimento)-1]='\0';

    verificaVazio(jogador->nome);
    verificaVazio(jogador->universidade);
    verificaVazio(jogador->cidadeNascimento);
    verificaVazio(jogador->estadoNascimento);

    return 8;
}

int main() {
    Jogador jogadores[3923];
    Fila fila;
    start(&fila);
    FILE* arquivo;
    arquivo = fopen("/tmp/players.csv", "r");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo.\n");
        return 1;
    }

    // Pular o cabeçalho e carregar todos os jogadores do arquivo
    char linha[1024];
    fgets(linha, sizeof(linha), arquivo);

    int totalJogadores = 0;
    while (lerJogador(arquivo, &jogadores[totalJogadores]) != EOF && totalJogadores < 3923) {
        totalJogadores++;
    }
    fclose(arquivo);

    // Ler os IDs da entrada padrão e inserir na fila
    char entrada[3923];
    while (scanf("%s", entrada) == 1 && strcmp(entrada, "FIM") != 0) {
        int idDesejado = atoi(entrada);

        for (int i = 0; i < totalJogadores; i++) {
            if (jogadores[i].id == idDesejado) {
                inserir(jogadores[i], &fila);

                double alturaTotal = 0;
                int cont = 0;
                Celula *temp = fila.primeiro;

                while (temp != NULL) {
                    alturaTotal += temp->jogador.altura;
                    cont++;
                    temp = temp->prox;
                }

                if (cont > 0) {
                    double mediaAltura = round(alturaTotal / cont);
                    printf("%d\n", (int)mediaAltura);
                }

                break;
            }
        }
    }

    char entrada3[250];
    int numRegistros;
    scanf("%d", &numRegistros);

    for (int i = 0; i < numRegistros; i++){
        scanf(" %[^\n]", entrada3);
        if(entrada3[0] == 'I'){
            int numero;
            sscanf(entrada3, "I %d", &numero);
            for (int i = 0; i < totalJogadores; i++) {
                if (jogadores[i].id == numero) {
                    inserir(jogadores[i], &fila);

                    double alturaTotal = 0;
                    int cont = 0;
                    Celula *temp = fila.primeiro;

                    while (temp != NULL) {
                        alturaTotal += temp->jogador.altura;
                        cont++;
                        temp = temp->prox;
                    }

                    if (cont > 0) {
                        double mediaAltura = round(alturaTotal / cont);
                        printf("%d\n", (int)mediaAltura);
                    }

                    break;
                }
            }
        } else if(entrada3[0] == 'R') {
            Jogador removido = remover(&fila);
            printf("(R) %s\n", removido.nome); 
        }
    }

    Celula *atual = fila.primeiro;
    while (atual != NULL) {
        imprimir(&atual->jogador); // Imprime todas as informações do jogador
        atual = atual->prox; // Move para o próximo elemento da fila
    }
   

    return 0;
}